xdsl/dialects/riscv.py

```diff
@@ -346,13 +351,55 @@ def parse(cls, parser: Parser) -> Self:
             regions=regions,
         )
 
+    @classmethod
+    def parse_args(cls, parser: Parser) -> Sequence[UnresolvedOperand]:
+        if arg := parser.parse_optional_unresolved_operand():
+            args = [arg]
+            while parser.parse_optional_punctuation(",") and (
+                arg := parser.parse_optional_unresolved_operand()
+            ):
+                args.append(arg)
+            return args
+        return []
+
+    @classmethod
+    def parse_attributes(cls, parser: Parser) -> Mapping[str, Attribute]:
+        return parser.parse_optional_attr_dict()
+
     def print(self, printer: Printer) -> None:
-        printer.print_operands(self.operands)
+        first = True
+        first &= not self.print_operands(printer, first)
+        processed_attrs = self.print_attributes(printer, first)
+        remaining_attrs = {
+            name: attr
+            for name, attr in self.attributes.items()
+            if name not in processed_attrs
+        }
+        printer.print_op_attributes(remaining_attrs)
         printer.print_regions(self.regions)
-        printer.print_op_attributes(self.attributes)
         printer.print(" : ")
         printer.print_operation_type(self)
 
+    def print_operands(self, printer: Printer, first: bool) -> Sequence[SSAValue]:
+        """
+        Print list of comma separated operands,
+        return the list of operands being processed.
+        """
+        if self.operands:
+            if not first:
+                printer.print(",")
+            printer.print(" ")
+            printer.print_list(self.operands, printer.print_operand)
+        return self.operands
+
+    def print_attributes(self, printer: Printer, first: bool) -> Set[str]:
+        """
+        Print list of attributes,
+        return the list of attributes being processed.
```

Comment:
what does this mean? can you please elaborate in the comment? it feels like this is another case of adding helper methods before it is clear that they will be used later on. Adding a `self.print_attributes` method that just forwards to priner.print_op_attributes makes the code less clear in the short term. Am I missing something? Is it necessary?

Reply:
I would suggest to first put up a pr that just drops the `()` around all operands. This should be a one-line change and does not yet need any of the helper functions. Let's avoid introducing all the helper functions, as these are controversial and I would like to discuss  them with Anton in person.

Reply:
Ah no I see it is being overridden below, then please expand on the comment on how the logic works. To me, it seems like the list of attributes returned is not "processed" but rather "printed", maybe "printed in the custom format". I'd recommend changing the name of the method to something like "print_known_attributes" or, "print_attributes_custom_format"